File: DAY-5/linearSearch_recursion.cpp
#include<bits/stdc++.h>
using namespace std;

bool find(int arr[],int index,int n,int target){
	if(index==n){
		return false;
	}

	return arr[index]==target || find(arr,++index,n,target);
}

int findIndex(int arr[],int index,int n,int target){
	if(index==n){
		return -1;
	}
	if(arr[index]==target){
		return index;
	}
	return findIndex(arr,++index,n,target);
}

int findIndexLast(int arr[],int index,int target){
	if(index==-1){
		return -1;
	}
	if(arr[index]==target){
		return index;
	}
	return findIndexLast(arr,--index,target);
}

vector<int> findAllIndex(int arr[],int n,int index,int target,vector<int> ans){
	if(index==n){
		return ans;
	}
	if(arr[index]==target){
		ans.push_back(index);
	}
	return findAllIndex(arr,n,++index,target,ans);	
}

int main(){
	int arr[]={1,23,23,23,23,23};
	int n=sizeof(arr)/sizeof(arr[0]);
	int target=23;
	vector<int> temp;

	cout<<(find(arr,0,n,target)?"True":"False")<<endl;
	
	cout<<findIndex(arr,0,n,target)<<endl;
	
	cout<<findIndexLast(arr,n-1,target)<<endl;
	
	vector<int> ans=findAllIndex(arr,n,0,target,temp);
	for(auto it: ans){
		cout<<it<<" ";
	}	
}

File: DAY-5/linearSearch_recursion.java
package com.CherishDSA;

import java.util.ArrayList;

public class linearSearch_recursion {
    public static boolean find(int[] arr,int target,int index){
        if(index==arr.length){
            return false;
        }
        return arr[index]==target || find(arr,target,++index);
    }
    public static int findIndex(int[] arr,int target,int index){
        if(index==arr.length){
            return -1;
        }
        if(arr[index]==target){
            return index;
        }
        return findIndex(arr,target,++index);
    }
    public static int findIndexLast(int[] arr,int target,int index){
        if(index<0){
            return -1;
        }
        if(arr[index]==target){
            return index;
        }
        return findIndex(arr,target,--index);
    }
    public static ArrayList findAllIndex(int[] arr,int target,int index, ArrayList<Integer> list){
        if(index==arr.length){
            return list;
        }
        if(arr[index]==target){
            list.add(index);
        }
        return findAllIndex(arr,target,++index,list);
    }
    static ArrayList findAllIndex2(int[] arr,int target,int index){
        ArrayList<Integer> list=new ArrayList<>();
        if(index==arr.length){
            return list;
        }
        if(arr[index]==target){
            list.add(index);
        }
        ArrayList<Integer> ansFromBelowCalls=findAllIndex2(arr,target,++index);
        list.addAll(ansFromBelowCalls);
        return list;
    }
    public static void main(String[] args) {
        int[] arr={23,567,45,23,567,567,456};
        int target=567;
//        System.out.println(find(arr,target,0));
//        System.out.println(findIndex(arr,target,0));
//        System.out.println(findIndexLast(arr,target,arr.length-1));
        System.out.println(findAllIndex2(arr,target,0));
    }
}
